### DISCUSSION TITLE
[Question] Question on the `wlp` of `while` command

### DISCUSSION THREAD

**Discussion by minseongg** [prosyslab-classroom/cs424-program-reasoning#317] [Q&A]
Hi, I have a question on the `wlp` of `while` command.

In the page 29 of [lecture 7](https://github.com/prosyslab-classroom/cs492-program-reasoning/blob/main/slides/lecture7.pdf), it says that the `wlp` of `while` command is as follows:

```
wlp(while E do S, Q) = I /\ (E /\ I -> wlp(S, I)) /\ (!E /\ I -> Q)
```

with assumption that an inductive invariant `I` is provided.

And I got confused when I applied this to the example in page 31.

```
while i > 1 {
  i := i - 2;
}
assert(i = 1);
```

For the above program, I used inductive invariant as `i >= 0` instead of `i >= 0 /\ odd(i)`.
(I think it is also inductive invariant because `wlp(i := i - 2, i >= 0) = (i >= 2)` and `i > 1 => i >= 2` is true.)

So in the above notation, `E` is `i > 1`, `S` is `i := i - 2`, `Q` is `i = 1` and `I` is `i >= 0`.

```
  I
= i >= 0

  E /\ I -> wlp(S, I)
= (i > 1) /\ (i >= 0) -> wlp(i := i - 2, i >= 0)
= (i > 1) /\ (i >= 0) -> (i - 2 >= 0)
= i > 1 -> i >= 2
= true

  !E /\ I -> Q
= !(i > 1) /\ (i >= 0) -> i = 1
= 0 <= i <= 1 -> i = 1
= !(0 <= i <= 1) \/ (i = 1)
= (i < 0 \/ i > 1) \/ (i = 1)
= i < 0 \/ i >= 1

  wlp(while E do S, Q)
= I /\ (E /\ I -> wlp(S, I)) /\ (!E /\ I -> Q)
= i >= 0 /\ true /\ (i < 0 \/ i >= 1)
= i >= 1
```

Therefore, `wlp(while i > 1 do i := i - 2, i = 1)` is `i >= 1` and so `{ i = 4 } while i > 1 do i := i - 2 { i = 1 }` is true because `i = 4 => i >= 1` is true. However, I think this is not true because the assertion fails when the initial value of `i` is 4. Is there anything I'm thinking wrong?

+) I searched for other references and found that there is a definition as follows ([Reference](https://www.lri.fr/~marche/MPRI-2-36-1/2013/poly1.pdf))

```
wlp(while E do S, Q) = I /\ (forall x1,...,xk (((E /\ I) -> wlp(S, I)) /\ ((!E /\ I) -> Q))[xi/wi])

where w1,...,wk is the set of assigned variables in statement S and x1,...,xk are fresh logic variables.
```

**Comment by KihongHeo** [comment]
Thanks a lot. That was my mistake.
Here is the problem.

The invariant you provide `I: i >=0` is not strong enough to show `!E /\ I -> Q`: `0 <= i <= 1 -> i = 1` is not true for all `i`.  The quantifier is omitted in the lecture slide. So the correct `wlp` is what you mentioned.

In short, the weakest precondition in the above example is `false`. This means that no possible precondition we can infer with the invariant `I: i >=0`.

**Comment by minseongg** [comment]
Thanks! It helped a lot.

**Comment by KihongHeo** [comment]
Since it is quite important, we will give you 1% bonus points. @Hhro plz take a memo.

**Comment by Hhro** [comment]
Ok, I'll do it right away.
